### PR TITLE
Add SPAKE2 password authenticated key exchange (RFC 9382)

### PR DESCRIPTION
**Pull request by randombit**

### CODE REVIEW

**Review by reneme — CHANGES_REQUESTED** [review]
I have a few suggestions re the API and some nits. Most of them, I prototyped [in another branch](https://github.com/reneme/botan/tree/jack/spake2), including an example.cpp.

File: src/lib/pake/spake2/spake2.h [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+      * where N/M are the output of hash to curve; this requires an attacker
+      * with a quantum computer to perform a discrete logarithm calculator per
+      * PAKE, rather than just once for the default (fixed) SPAKE2 parameters.
+      *
+      * Here the shared secret a random scalar. It should have been generated
+      * using a memory hard function such as Argon2id.
+      *
+      * # ⚠️  Warning
+      *
+      * This interface exists primarily for testing, and is not safe for general use.
```
[the remark this answers is not in the document]
Comment (randombit) [reply src/lib/pake/spake2/spake2.h:91]: It does allow this (or Argon2id with different parameters, etc), so there is utility outside of the testing scenario. However it seems like an obvious footgun already, and a `span` argument instead seems even more prone to misuse.

I suppose I can amend the comment to make it clear it is usable as long as the scalar is derived from the password using a MHF and with care to avoid bias, referring to the relevant RFC sections.

**Comment by randombit** [comment]
@reneme Finally coming back to this. I do like your approach in terms of footgun resistance but I think it does not work for FFI where we must put the state into a `unique_ptr` so we can return it out to the FFI caller.

File: src/lib/pake/spake2/spake2.h [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+         }
+      }
+
+      const EC_AffinePoint& spake2_their_pt(SPAKE2_PeerId whoami) const {
+         if(whoami == SPAKE2_PeerId::PeerA) {
+            return m_params.second;  // N
+         } else {
+            return m_params.first;  // M
+         }
+      }
```
[the remark this answers is not in the document]
Comment (randombit) [reply src/lib/pake/spake2/spake2.h]: Done

**Comment by randombit** [comment]
Still todo

- [ ] More tests especially negative tests
- [ ] Benchmarks
- [ ] FFI + FFI tests + Python

**Comment by reneme** [comment]
> I do like your approach in terms of footgun resistance but I think it does not work for FFI where we must put the state into a unique_ptr so we can return it out to the FFI caller.

The `State` object is stack-based and can't be copied. Therefore it is a bit awkward to use with `std::unique_ptr`, for sure. But I wouldn't see that as a show stopper. A small wrapper with a member of `State` would be enough to mitigate this. If you really want, you could even just use `std::optional<>` as the wrapper type...

```C++
std::unique_ptr<std::optional<State>> wrapper;
State s1{}; // from the PAKE
wrapper.reset(new std::optional<State>(std::move(s1)));
// ---- FFI API boundary
State s2 = std::move(wrapper->value());
```
[Godbolt.org](https://godbolt.org/z/e9v8WbMYj)

**Comment by randombit** [comment]
I think we're going to go for SPAKE2+ instead

**Comment by randombit** [comment]
SPAKE2+ #5711
